db/revision.go

```diff
@@ -229,25 +229,28 @@ func (body Body) getExpiry() (uint32, bool, error) {
 	return *expiry, true, err
 }
 
-// nonJSONPrefix is used to ensure old revision bodies aren't hidden from N1QL/Views.
-const nonJSONPrefix = byte(1)
-
-// Looks up the raw JSON data of a revision that's been archived to a separate doc.
+// getOldRevisionJSON looks up the raw JSON data of a revision that's been archived to a separate doc.
 // If the revision isn't found (e.g. has been deleted by compaction) returns 404 error.
-func (c *DatabaseCollection) getOldRevisionJSON(ctx context.Context, docid string, rev string) ([]byte, error) {
-	data, _, err := c.dataStore.GetRaw(oldRevisionKey(docid, rev))
+func (c *DatabaseCollection) getOldRevisionJSON(ctx context.Context, docid string, revOrCV string) ([]byte, error) {
+	data, _, err := c.dataStore.GetRaw(oldRevisionKey(docid, revOrCV))
 	if base.IsDocNotFoundError(err) {
-		base.DebugfCtx(ctx, base.KeyCRUD, "No old revision %q / %q", base.UD(docid), rev)
-		err = ErrMissing
-	}
-	if data != nil {
-		// Strip out the non-JSON prefix
-		if len(data) > 0 && data[0] == nonJSONPrefix {
-			data = data[1:]
-		}
-		base.DebugfCtx(ctx, base.KeyCRUD, "Got old revision %q / %q --> %d bytes", base.UD(docid), rev, len(data))
+		base.DebugfCtx(ctx, base.KeyCRUD, "No old revision %q / %q", base.UD(docid), revOrCV)
+		return nil, ErrMissing
+	} else if err != nil {
+		return nil, err
+	}
+	kind, revData := stripNonJSONPrefix(data)
+	switch kind {
+	case nonJSONPrefixKindRevBody:
+		base.DebugfCtx(ctx, base.KeyCRUD, "Got old revision %q / %q --> %d bytes", base.UD(docid), revOrCV, len(revData))
+		return revData, nil
+	case nonJSONPrefixKindRevPtr:
+		// pointer to a CV-keyed backup revision - do another fetch by CV for the body
+		base.DebugfCtx(ctx, base.KeyCRUD, "Found old revision pointer %q -> %q / %q", revOrCV, base.UD(docid), string(revData))
+		return c.getOldRevisionJSON(ctx, docid, string(revData))
+	default:
+		return nil, fmt.Errorf("unexpected prefix %b for old revision doc %q / %q", kind, base.UD(docid).String(), revOrCV)
```

Comment:
Do you want to do, or are these equivalent?


```suggestion
		return nil, base.RedactErrorf("unexpected prefix %b for old revision doc %q / %q", kind, base.UD(docid).String(), revOrCV)
```

Reply:
They're mostly equivalent here. The only difference that can occur when we use `RedactErrorf` is if we're using the error in a returned HTTP response, where the response gets the unredacted version, and the logging includes the redacted version.

In this case it _can_ do that, but it's not a huge deal and I don't expect this default case to be hit anyway.